src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs

```diff
@@ -1384,5 +1393,526 @@ void M(bool b)
+                // (9,32): error CS0453: The type 'T' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'E<T>'
+                //     void M4<T>(delegate*<E<T>> ptr) {}
+                Diagnostic(ErrorCode.ERR_ValConstraintNotSatisfied, "ptr").WithArguments("E<T>", "T", "T").WithLocation(9, 32)
+            );
+        }
+
+        [Fact]
+        public void NewFunctionPointerType()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    void M()
+    {
+        _ = new delegate*<void>();
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (6,13): error CS1919: Unsafe type 'delegate*<void>' cannot be used in object creation
+                //         _ = new delegate*<void>();
+                Diagnostic(ErrorCode.ERR_UnsafeTypeInObjectCreation, "new delegate*<void>()").WithArguments("delegate*<void>").WithLocation(6, 13)
+            );
+        }
+
+        [Fact]
+        public void IndexerAccessOnFunctionPointer()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    void M(delegate*<void> ptr)
+    {
+        _ = ptr[0];
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (6,13): error CS0021: Cannot apply indexing with [] to an expression of type 'delegate*<void>'
+                //         _ = ptr[0];
+                Diagnostic(ErrorCode.ERR_BadIndexLHS, "ptr[0]").WithArguments("delegate*<void>").WithLocation(6, 13)
+            );
+        }
+
+        [Fact]
+        public void ClsComplaince()
+        {
+
+            var comp = CreateCompilationWithFunctionPointers(@"
+using System;
+[assembly: CLSCompliant(true)]
+[CLSCompliant(true)]
+public class C
+{
+    private unsafe void M1(delegate*<void> m) {}
+    internal unsafe void M2(delegate*<void> m) {}
+    public unsafe void M3(delegate*<void> m) {}
+}");
+
+            comp.VerifyDiagnostics(
+                // (9,43): warning CS3001: Argument type 'delegate*<void>' is not CLS-compliant
+                //     public unsafe void M3(delegate*<void> m) {}
+                Diagnostic(ErrorCode.WRN_CLS_BadArgType, "m").WithArguments("delegate*<void>").WithLocation(9, 43)
+            );
+        }
+
+        [Fact]
+        public void CannotMakeFunctionPointerConst()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    const delegate*<void> field = null;
+    public static void M()
+    {
+        const delegate*<void> local = null;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (4,5): error CS0283: The type 'delegate*<void>' cannot be declared const
+                //     const delegate*<void> field = null;
+                Diagnostic(ErrorCode.ERR_BadConstType, "const").WithArguments("delegate*<void>").WithLocation(4, 5),
+                // (4,35): error CS0133: The expression being assigned to 'C.field' must be constant
```

Comment:
>The expression being assigned to 'C.field' must be constant [](start = 41, length = 59)

Is the expression not constant?

Reply:
This is likely an error message we could suppress for this case, but technically no, since `null` is subject to an implicit conversion here. We report the same error messages today for `int*` consts and family.

---
In reply to: [408582206](https://github.com/dotnet/roslyn/pull/43248#discussion_r408582206) [](ancestors = 408582206)